**Note to grader:** Each question consists of parts, e.g. Q1(i), Q1(ii), etc. Each part must be first graded  on a 0-4 scale, following the standard NJIT convention (A:4, B+: 3.5, B:3, C+: 2.5, C: 2, D:1, F:0). However, any given item may be worth 4 or 8 points; if an item is worth 8 points, you need to accordingly scale the 0-4 grade.


The total score must be re-scaled to 100. That should apply to all future assignments so that Canvas assigns the same weight on all assignments. 



# Assignment 1



## Preparation Steps

In [1]:
# Import all necessary python packages
import numpy as np
import torch



## <font color = 'blue'> Question 1. Basic Operations with Tensors </font>

Your task for this question is to follow the NumPy  [**tutorial**](https://jalammar.github.io/visual-numpy/?fbclid=IwAR0tSntx5mj1aHteokRKrT4G6z77M3z0Quj40AQZ9mvKlhs2RTN3xXrc6Eo) and 'mirror' each of the operations presented in the tutorial with tensors in PyTorch. 

You may find useful to consult this PyTorch introductory [tutorial](https://jhui.github.io/2018/02/09/PyTorch-Basic-operations/), and as always the full PyTorch [documentation](https://pytorch.org/docs/stable/torch.html) is the ultimate resource.

*(Please insert cells below for your answers )*

---
#### Creating Arrays

In [4]:
np.array([1,2,3])

array([1, 2, 3])

Use to inistalize an array's shape

In [5]:
np.ones(3)
np.zeros(3)
np.random.random(3)

array([0.57561424, 0.3192465 , 0.31137391])

#### Array Arithmetic
Create two arrays and then lets do some simple arithmetic.

In [10]:
data = np.array([1,2])
ones = np.ones(2)

#Keep in mind that the arrays must be of the same shape for this orperation to work.
sum = data + ones 

sum


array([2., 3.])

In [12]:
difference = data - ones
product = data * ones 
quotient = data / ones 

print(difference,product,quotient)

[0. 1.] [1. 2.] [1. 2.]


This can also be done between a vector and a scalar value.<br>
This is called "Broadcasting"

In [15]:
data = np.array([1,2])

scalar_transforms = data * 1.6 
scalar_transforms

array([1.6, 3.2])

#### Indexing Numpy Arrays
Same as indexing python lists

In [16]:
data = np.array([1,2,3])
print(data[0],data[1],data[0:2],data[1:])

1 2 [1 2] [2 3]


#### Aggregation 
Many more operations than listed here

In [20]:
data = np.array([1,2,3])
max = data.max()
min = data.min() 
sum = data.sum()
mean = data.mean() 
product = data.prod() 
sd = data.std()

print(f"Max: {max}, Min: {min}, Mean: {mean}, Sum: {sum}, Product: {product}, Standard Deviation: {sd}")

Max: 3, Min: 1, Mean: 2.0, Sum: 6, Product: 6, Standard Deviation: 0.816496580927726


#### In More Dimensions 
(Arrays with more and 1 dimension)

In [28]:
#Two dimension Matrix
# Number of brackets indicates how many dimensions
np.array([[1,2],[3,4]])

array([[1, 2],
       [3, 4]])

In [26]:
#((row,column))
# Number of parenthesis indicates how many dimensions
np.ones((3,2))
np.zeros((3,2))
np.random.random((3,2))

array([[8.90811670e-01, 3.70601928e-01],
       [5.15441824e-01, 8.63644570e-01],
       [6.94282565e-04, 7.97289831e-01]])

#### Matrix Arithmetic

In [30]:
ones = np.ones((2,2))
data = np.array([[1,2],[3,4]])

sum = ones + data

sum

array([[2., 3.],
       [4., 5.]])

*this can be done with all of the arithmetic operators (+-*/)

We can get away with doing these arithmetic operations on matrices of different size only if the different dimension is one (e.g. the matrix has only one column or one row)

In [36]:
data = np.array([[1,2],[3,4],[5,6]])
ones_row = np.ones(2)

sum = data + ones_row

print(f"data + ones_row = \n{data} \n+ \n{ones_row} \n= \n{sum}")

data + ones_row = 
[[1 2]
 [3 4]
 [5 6]] 
+ 
[1. 1.] 
= 
[[2. 3.]
 [4. 5.]
 [6. 7.]]


#### Dot Product 
A key distinction to make with arithmetic is the case of matrix multiplication using the dot product. NumPy gives every matrix a dot() method we can use to carry-out dot product operations with other matrices:<br>
**Matrices must have one common dimension!**

In [47]:
data = np.array([1,2,3])
powers_of_ten = np.array([[1,10],[100,1000],[10000,100000]])

dot_prod = data.dot(powers_of_ten) 
dot_prod

array([ 30201, 302010])

>What is happening: 1x1 + 2x100 + 3x10,000 | 1x10 + 2x1,000 + 3x100,000 = 30201 | 302010

#### Matrix Indexing 
Indexing and slicing operations become even more useful when we’re manipulating matrices:<br>
**array[Row,Column]**

In [51]:
data = np.array([[1,2],[3,4],[5,6]])

#[Row,Column]- If no Row or Column value is entered it exp
ex1 = data[0,1] 

#Slicing
ex2 = data[1:3]
ex3 = data[0:2,0]


print(f"ex1 = {ex1}\nex2 = {ex2}\nex3 = {ex3}")

ex1 = 2
ex2 = [[3 4]
 [5 6]]
ex3 = [1 3]


#### Matrix Aggregation
We can aggregate matrices the same way we aggregated vectors:<br>
^ This can be done with alot of other aggregation methods:)

In [53]:
data = np.array([[1,2],[3,4],[5,6]])

max = data.max()
min = data.min()
sum = data.sum()

print(f"max = {max}, min = {min}, sum = {sum} ")

max = 6, min = 1, sum = 21 


Not only can we aggregate all the values in a matrix, but we can also aggregate across the rows or columns by using the axis parameter:


In [56]:
max_axis_0 = data.max(axis=0)
max_axis_1 = data.max(axis=1)

print(f"Axis 0 max: {max_axis_0}, Axis 1 max: {max_axis_1}")


Axis 0 max: [5 6], Axis 1 max: [2 4 6]


#### Transposing and Reshaping
A common need when dealing with matrices is the need to rotate them. This is often the case when we need to take the dot product of two matrices and need to align the dimension they share. NumPy arrays have a convenient property called **T** to get the transpose of a matrix:<br>

In [58]:
data = np.array([[1,2],[3,4],[5,6]])

t_data = data.T 

print(f"Data:\n{data}\nTransposed Data:\n{t_data}")

Data:
[[1 2]
 [3 4]
 [5 6]]
Transposed Data:
[[1 3 5]
 [2 4 6]]


In more advanced use case, you may find yourself needing to switch the dimensions of a certain matrix. This is often the case in machine learning applications where a certain model expects a certain shape for the inputs that is different from your dataset. NumPy’s **reshape()** method is useful in these cases. You just pass it the new dimensions you want for the matrix. You can pass -1 for a dimension and NumPy can infer the correct dimension based on your matrix:

In [69]:
data = np.array([1,2,3,4,5,6])

#Play with this:
reshape_data = data.reshape(3,2)

print(f"Data: {data}\nReshaped Data:\n{reshape_data} ")

Data: [1 2 3 4 5 6]
Reshaped Data:
[[1 2]
 [3 4]
 [5 6]] 


#### Yet More Dimensions
NumPy can do everything we’ve mentioned in any number of dimensions. Its central data structure is called ndarray (N-Dimensional Array) for a reason.


In [71]:
n_data = np.array([[[1,2],[3,4]],[[5,6],[7,8]]]) 
n_data

array([[[1, 2],
        [3, 4]],

       [[5, 6],
        [7, 8]]])

In [78]:
n_ones = np.ones((4,3,2))
n_zeros = np.zeros((4,3,2))
n_random = np.random.random((4,3,2))

print(f"n_ones:\n{n_ones}")

n_ones:
[[[1. 1.]
  [1. 1.]
  [1. 1.]]

 [[1. 1.]
  [1. 1.]
  [1. 1.]]

 [[1. 1.]
  [1. 1.]
  [1. 1.]]

 [[1. 1.]
  [1. 1.]
  [1. 1.]]]


##### np.ones((4,3,2)): 4 High, 3 Wide, 2 depth or y = 4, x = 3, z = 2 in 3D Space
* Note: Keep in mind that when you print a 3-dimensional NumPy array, the text output visualizes the array differently than shown here. NumPy’s order for printing n-dimensional arrays is that the last axis is looped over the fastest, while the first is the slowest.

#### Practical Usage<br>
##### Formulas<br>
Implementing mathematical formulas that work on matrices and vectors is a key use case to consider NumPy for. It’s why NumPy is the darling of the scientific python community. For example, consider the mean square error formula that is central to supervised machine learning models tackling regression problems:




In [ ]:
# For grader use only

G = [0]*2


# insert grade here  (from 0 to 8)
# G[1] = 

# please justify point subtraction  s

##  <font color = 'blue'> Question 2. Quadratic Regression

In the lecture we discussed a simple regression problem, where points are coming from the line $y = 2x + 1$, plus some noise.  For this question you are asked to: 

(i) Generate data points coming from a quadratic function: $ y = -x^2 + 3x +10 $, plus some noise. <br>
(ii) Modify the PyTorch model from the class in order for it to learn a quadratic function of the form $y = a x^2 + bx +c$. <br>
(iii) Train your model and report what values it computes. (Sanity check: These sould be close to -1, 3, 10)




In [ ]:
# for grader use only

# insert grade here  
# part (i): 4, part(ii) 8, part (iii) 8

# G[2] = 
#
# please justify point subtractions

In [ ]:
# total score
max_score = 36
final_score = sum(G)*(100/max_score)